File: app/Sources/Lippy/Separator.swift
import AppKit
import ApplicationServices

/// Decides whether inserted text needs a space in front of it.
///
/// Dictation arrives one utterance at a time and is pasted at the cursor, which
/// knows nothing about what is already there. Two sentences in a row therefore
/// run together: "Ship it on Tuesday.The bug is fixed."
///
/// Always prepending a space is not the fix. It leaves a stray space in every
/// empty field, at the start of every line, and after every opening bracket. The
/// decision has to be made from what actually precedes the insertion point.
enum Separator {

    /// Characters after which a space would be wrong.
    private static let openers: Set<Character> = ["(", "[", "{", "<", "\"", "'", "\u{201C}", "\u{2018}", "/", "-", "\u{2013}", "\u{2014}", "@", "#", "$"]

    /// Pure decision, so it can be tested without a running app.
    ///
    /// - Parameters:
    ///   - preceding: the character immediately before the cursor, or nil when
    ///     the field is empty or the app does not expose its contents.
    ///   - inserting: the text about to be pasted.
    static func needed(after preceding: Character?, inserting: String) -> Bool {
        guard let preceding else { return false }
        guard let first = inserting.first else { return false }

        // Already separated, or at the start of a line.
        if preceding.isWhitespace || preceding.isNewline { return false }
        // A space after an opening bracket or quote is wrong.
        if openers.contains(preceding) { return false }
        // Text that opens with its own punctuation supplies its own spacing.
        if first.isWhitespace || first.isNewline { return false }
        if first.isPunctuation, !openers.contains(first) { return false }

        return true
    }

    /// Reads the character immediately before the insertion point in the focused
    /// field. Returns nil when nothing has focus, when the app does not publish
    /// its text, or when the cursor is at the very start.
    static func characterBeforeCursor() -> Character? {
        guard AXIsProcessTrusted() else { return nil }

        var focused: CFTypeRef?
        guard AXUIElementCopyAttributeValue(
                AXUIElementCreateSystemWide(),
                kAXFocusedUIElementAttribute as CFString, &focused) == .success,
              let raw = focused else { return nil }
        let element = raw as! AXUIElement

        var valueRef: CFTypeRef?
        guard AXUIElementCopyAttributeValue(
                element, kAXValueAttribute as CFString, &valueRef) == .success,
              let text = valueRef as? String, !text.isEmpty else { return nil }

        var rangeRef: CFTypeRef?
        guard AXUIElementCopyAttributeValue(
                element, kAXSelectedTextRangeAttribute as CFString, &rangeRef) == .success
        else { return nil }

        var range = CFRange()
        guard AXValueGetValue(rangeRef as! AXValue, .cfRange, &range) else { return nil }

        // Accessibility ranges are UTF-16 offsets, so the string has to be
        // indexed the same way. Using Character offsets would land in the wrong
        // place after any emoji or combining mark.
        let units = Array(text.utf16)
        let location = range.location
        guard location > 0, location <= units.count else { return nil }
        guard let scalar = Unicode.Scalar(units[location - 1]) else { return nil }
        return Character(scalar)
    }

    /// The text to paste, with a leading space when one is needed.
    ///
    /// Fails closed: when the focused app does not publish its contents, no
    /// space is added. A missing space between two utterances is a visible
    /// nuisance the user can fix in one keystroke. A spurious leading space
    /// appears on the *first* dictation into every such app, which is worse.
    static func prepare(_ text: String) -> String {
        needed(after: characterBeforeCursor(), inserting: text) ? " " + text : text
    }
}

// MARK: - Self test

extension Separator {
    /// `Lippy --selftest-separator`. Needs no daemon, no permissions and no
    /// focused window, so it runs anywhere including CI.
    static func runSelfTest() -> Int32 {
        // (preceding, inserting, expected, why)
        let cases: [(Character?, String, Bool, String)] = [
            (nil, "Hello", false, "empty field, or an app that does not publish its text"),
            (".", "How are you", true, "the reported bug: sentence hard against the last full stop"),
            ("d", "How are you", true, "mid-sentence continuation still needs a space"),
            ("?", "Yes", true, "after a question mark"),
            (" ", "Hello", false, "already separated"),
            ("\n", "Hello", false, "start of a line"),
            ("\t", "Hello", false, "after a tab"),
            ("(", "Hello", false, "opening bracket"),
            ("\"", "Hello", false, "opening quote"),
            ("-", "Hello", false, "hyphen, likely mid-word"),
            ("/", "Hello", false, "path or URL"),
            ("@", "Hello", false, "handle or address"),
            (".", ", and then", false, "inserted text opens with its own punctuation"),
            (".", " already spaced", false, "inserted text supplies its own space"),
            (".", "\"Quoted\"", true, "an opening quote is text, not punctuation to hug"),
            (".", "", false, "nothing to insert"),
        ]

        var failures = 0
        for (preceding, inserting, expected, why) in cases {
            let got = needed(after: preceding, inserting: inserting)
            let ok = got == expected
            if !ok { failures += 1 }
            let shown = preceding.map { String($0).debugDescription } ?? "nil"
            print("  \(ok ? "PASS" : "FAIL")  after \(shown) inserting \(inserting.debugDescription) -> \(got)  (\(why))")
        }
        print("\n  \(cases.count - failures)/\(cases.count) passed")
        return failures == 0 ? 0 : 1
    }
}

File: daemon/separator.py
"""Whether inserted text needs a space in front of it.

Ported from `needed(after:inserting:)` in
`app/Sources/Lippy/Separator.swift`, which is pure and ports as written. The
reading of the character before the cursor does not port, because that is an
accessibility call on one platform and a UI Automation call on the other, and
neither belongs in a decision.

Dictation arrives one utterance at a time and is pasted at the cursor, which
knows nothing about what is already there. Two sentences in a row therefore run
together: "Ship it on Tuesday.The bug is fixed."

Always prepending a space is not the fix. It leaves a stray space in every empty
field, at the start of every line, and after every opening bracket. The decision
has to be made from what actually precedes the insertion point.
"""

from __future__ import annotations

import unicodedata

# Characters after which a space would be wrong. Written as escapes, the way
# the Swift is, so that a reader can tell the curly quotes and the two long
# dashes apart on any screen.
OPENERS = frozenset(
    [
        "(", "[", "{", "<",
        '"', "'", "\u201C", "\u2018",
        "/", "-", "\u2013", "\u2014",
        "@", "#", "$",
    ]
)


def _is_punctuation(character: str) -> bool:
    # Unicode general category P, which is what Swift's isPunctuation reports.
    # Note what this excludes: "$" is a currency symbol and "<" is a maths
    # symbol, so neither is punctuation here. Both are in OPENERS anyway, which
    # is the only reason that distinction never surfaces.
    return unicodedata.category(character).startswith("P")


def needed(after: str | None, inserting: str) -> bool:
    """True when a leading space should be added to `inserting`.

    `after` is the single character immediately before the cursor, or None when
    the field is empty, nothing has focus, or the application does not publish
    its contents. A string of any other length is a caller that has passed the
    whole field, which Swift's type signature made impossible and Python's does
    not.
    """
    if not isinstance(inserting, str):
        raise TypeError(f"inserting must be a str, got {inserting!r}")
    if after is None:
        return False
    if not isinstance(after, str):
        raise TypeError(f"after must be a single character or None, got {after!r}")
    if len(after) != 1:
        raise ValueError(
            f"after must be exactly one character or None, got {after!r}")

    if not inserting:
        return False
    first = inserting[0]

    # Already separated, or at the start of a line.
    if after.isspace():
        return False
    # A space after an opening bracket or quote is wrong.
    if after in OPENERS:
        return False
    # Text that opens with its own punctuation supplies its own spacing.
    if first.isspace():
        return False
    if _is_punctuation(first) and first not in OPENERS:
        return False

    return True


def prepare(text: str, preceding: str | None) -> str:
    """The text to paste, with a leading space when one is needed.

    Fails closed. When the focused application does not publish its contents,
    `preceding` is None and no space is added. A missing space between two
    utterances is a visible nuisance the user fixes in one keystroke. A spurious
    leading space appears on the *first* dictation into every such application,
    which is worse.
    """
    return " " + text if needed(preceding, text) else text
